File: connect/src/main/java/org/transdroid/connect/clients/rtorrent/Rtorrent.java
package org.transdroid.connect.clients.rtorrent;

import com.jakewharton.retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;

import org.reactivestreams.Publisher;
import org.transdroid.connect.Configuration;
import org.transdroid.connect.clients.Feature;
import org.transdroid.connect.model.Torrent;
import org.transdroid.connect.model.TorrentStatus;
import org.transdroid.connect.util.OkHttpBuilder;
import org.transdroid.connect.util.RxUtil;

import java.io.InputStream;
import java.util.Date;

import io.reactivex.Flowable;
import io.reactivex.FlowableTransformer;
import io.reactivex.functions.Function;
import nl.nl2312.xmlrpc.Nothing;
import nl.nl2312.xmlrpc.XmlRpcConverterFactory;
import retrofit2.Retrofit;

public final class Rtorrent implements
		Feature.Version,
		Feature.Listing,
		Feature.StartingStopping,
		Feature.ResumingPausing,
		Feature.AddByFile,
		Feature.AddByUrl,
		Feature.AddByMagnet {

	private final Configuration configuration;
	private final Service service;

	public Rtorrent(Configuration configuration) {
		this.configuration = configuration;
		Retrofit retrofit = new Retrofit.Builder()
				.baseUrl(configuration.baseUrl())
				.client(new OkHttpBuilder(configuration).build())
				.addCallAdapterFactory(RxJava2CallAdapterFactory.create())
				.addConverterFactory(XmlRpcConverterFactory.create())
				.build();
		this.service = retrofit.create(Service.class);
	}

	@Override
	public Flowable<String> clientVersion() {
		return service.clientVersion(configuration.endpoint(), Nothing.NOTHING)
				.cache(); // Cached, as it is often used but 'never' changes
	}

	@Override
	public Flowable<Torrent> torrents() {
		return service.torrents(
				configuration.endpoint(),
				"",
				"main",
				"d.hash=",
				"d.name=",
				"d.state=",
				"d.down.rate=",
				"d.up.rate=",
				"d.peers_connected=",
				"d.peers_not_connected=",
				"d.bytes_done=",
				"d.up.total=",
				"d.size_bytes=",
				"d.left_bytes=",
				"d.creation_date=",
				"d.complete=",
				"d.is_active=",
				"d.is_hash_checking=",
				"d.base_path=",
				"d.base_filename=",
				"d.message=",
				"d.custom=addtime",
				"d.custom=seedingtime",
				"d.custom1=",
				"d.peers_complete=",
				"d.peers_accounted=")
				.compose(RxUtil.<TorrentSpec>asList())
				.map(new Function<TorrentSpec, Torrent>() {
					@Override
					public Torrent apply(TorrentSpec torrentSpec) throws Exception {
						return new Torrent(
								torrentSpec.hash.hashCode(),
								torrentSpec.hash,
								torrentSpec.name,
								torrentStatus(torrentSpec.state, torrentSpec.isComplete, torrentSpec.isActive, torrentSpec.isHashChecking),
								torrentSpec.basePath.substring(0, torrentSpec.basePath.indexOf(torrentSpec.baseFilename)),
								(int) torrentSpec.downloadRate,
								(int) torrentSpec.uploadRate,
								(int) torrentSpec.seedersConnected,
								(int) (torrentSpec.peersConnected + torrentSpec.peersNotConnected),
								(int) torrentSpec.leechersConnected,
								(int) (torrentSpec.peersConnected + torrentSpec.peersNotConnected),
								torrentSpec.downloadRate > 0 ? (torrentSpec.bytesleft / torrentSpec.downloadRate) : Torrent.UNKNOWN,
								torrentSpec.bytesDone,
								torrentSpec.bytesUploaded,
								torrentSpec.bytesTotal,
								torrentSpec.bytesDone / torrentSpec.bytesTotal,
								0F,
								torrentSpec.label,
								torrentTimeAdded(torrentSpec.timeAdded, torrentSpec.timeCreated),
								torrentTimeFinished(torrentSpec.timeFinished),
								torrentSpec.errorMessage
						);
					}
				});
	}

	@Override
	public Flowable<Torrent> start(final Torrent torrent) {
		return service.start(
				configuration.endpoint(),
				torrent.hash()).map(new Function<Void, Torrent>() {
			@Override
			public Torrent apply(Void result) throws Exception {
				return torrent.mimicStart();
			}
		});
	}

	@Override
	public Flowable<Torrent> stop(final Torrent torrent) {
		return service.stop(
				configuration.endpoint(),
				torrent.hash()).map(new Function<Void, Torrent>() {
			@Override
			public Torrent apply(Void result) throws Exception {
				return torrent.mimicStart();
			}
		});
	}

	@Override
	public Flowable<Void> addByFile(InputStream file) {
		// TODO
		return null;
	}

	@Override
	public Flowable<Void> addByUrl(final String url) {
		return clientVersion().compose(clientVersionAsInt).flatMap(new Function<Integer, Publisher<Integer>>() {
			@Override
			public Publisher<Integer> apply(Integer integer) throws Exception {
				if (integer > 904) {
					return service.loadStart(
							configuration.endpoint(),
							"",
							url);
				} else {
					return service.loadStart(
							configuration.endpoint(),
							url);
				}
			}
		}).map(new Function<Integer, Void>() {
			@Override
			public Void apply(Integer integer) throws Exception {
				return null;
			}
		});
	}

	@Override
	public Flowable<Void> addByMagnet(final String magnet) {
		return clientVersion().compose(clientVersionAsInt).flatMap(new Function<Integer, Publisher<Integer>>() {
			@Override
			public Publisher<Integer> apply(Integer integer) throws Exception {
				if (integer > 904) {
					return service.loadStart(
							configuration.endpoint(),
							"",
							magnet);
				} else {
					return service.loadStart(
							configuration.endpoint(),
							magnet);
				}
			}
		}).map(new Function<Integer, Void>() {
			@Override
			public Void apply(Integer integer) throws Exception {
				return null;
			}
		});
	}

	private FlowableTransformer<String, Integer> clientVersionAsInt = new FlowableTransformer<String, Integer>() {
		@Override
		public Publisher<Integer> apply(Flowable<String> version) {
			return version.map(new Function<String, Integer>() {
				@Override
				public Integer apply(String version) throws Exception {
					if (version == null)
						return 10000;
					try {
						String[] versionParts = version.split("\\.");
						return (Integer.parseInt(versionParts[0]) * 10000) + (Integer.parseInt(versionParts[1]) * 100) + Integer.parseInt
								(versionParts[2]);
					} catch (NumberFormatException e) {
						return 10000;
					}
				}
			});
		}
	};

	private TorrentStatus torrentStatus(long state, long complete, long active, long checking) {
		if (state == 0) {
			return TorrentStatus.QUEUED;
		} else if (active == 1) {
			if (complete == 1) {
				return TorrentStatus.SEEDING;
			} else {
				return TorrentStatus.DOWNLOADING;
			}
		} else if (checking == 1) {
			return TorrentStatus.CHECKING;
		} else {
			return TorrentStatus.PAUSED;
		}
	}

	private Date torrentTimeAdded(String timeAdded, long timeCreated) {
		if (timeAdded != null || timeAdded.trim().length() != 0) {
			return new Date(Long.parseLong(timeAdded.trim()) * 1000L);
		}
		return new Date(timeCreated * 1000L);
	}

	private Date torrentTimeFinished(String timeFinished) {
		if (timeFinished == null || timeFinished.trim().length() == 0)
			return null;
		return new Date(Long.parseLong(timeFinished.trim()) * 1000L);
	}

}

File: connect/src/main/java/org/transdroid/connect/clients/rtorrent/Rtorrent.kt
package org.transdroid.connect.clients.rtorrent

import io.reactivex.Completable
import io.reactivex.Flowable
import io.reactivex.Single
import io.reactivex.functions.BiFunction
import nl.nl2312.xmlrpc.XmlRpcConverterFactory
import org.transdroid.connect.Configuration
import org.transdroid.connect.clients.Feature
import org.transdroid.connect.model.*
import org.transdroid.connect.util.OkHttpBuilder
import org.transdroid.connect.util.flatten
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import java.io.InputStream
import java.util.*

class Rtorrent(private val configuration: Configuration) :
        Feature.Version,
        Feature.Listing,
        Feature.Details,
        Feature.StartingStopping,
        Feature.ResumingPausing,
        Feature.AddByFile,
        Feature.AddByUrl,
        Feature.AddByMagnet {

    private val xmlrpcSizeMinimum = 2 * 1024 * 1024
    private val xmlrpcSizePadding = 1280

    private val service: Service = Retrofit.Builder()
            .baseUrl(configuration.baseUrl)
            .client(OkHttpBuilder.build(configuration))
            .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
            .addConverterFactory(XmlRpcConverterFactory.builder()
                    .addArrayDeserializer(TorrentSpec::class.java) { arrayValues ->
                        TorrentSpec(
                                arrayValues.asString(0),
                                arrayValues.asString(1),
                                arrayValues.asLong(2),
                                arrayValues.asLong(3),
                                arrayValues.asLong(4),
                                arrayValues.asLong(5),
                                arrayValues.asLong(6),
                                arrayValues.asLong(7),
                                arrayValues.asLong(8),
                                arrayValues.asLong(9),
                                arrayValues.asLong(10),
                                arrayValues.asLong(11),
                                arrayValues.asLong(12),
                                arrayValues.asLong(13),
                                arrayValues.asLong(14),
                                arrayValues.asString(15),
                                arrayValues.asString(16),
                                arrayValues.asString(17),
                                arrayValues.asString(18),
                                arrayValues.asString(19),
                                arrayValues.asString(20),
                                arrayValues.asLong(21),
                                arrayValues.asLong(22)

                        )
                    }
                    .addArrayDeserializer(TrackerSpec::class.java) { arrayValues ->
                        TrackerSpec(arrayValues.asString(0))
                    }
                    .addArrayDeserializer(FileSpec::class.java) { arrayValues ->
                        FileSpec(
                                arrayValues.asString(0),
                                arrayValues.asLong(1),
                                arrayValues.asLong(2),
                                arrayValues.asLong(3),
                                arrayValues.asLong(4),
                                arrayValues.asString(5)
                        )
                    }
                    .create())
            .build().create(Service::class.java)

    override fun clientVersion(): Single<String> {
        return service.clientVersion(configuration.endpoint)
                .cache() // Cached, as it is often used but 'never' changes
    }

    private fun Single<String>.asVersionInt(): Single<Int> {
        return this.map {
            if (it == null) 10000 else {
                val versionParts = it.split(".")
                versionParts[0].toInt() * 10000 + versionParts[1].toInt() * 100 + versionParts[2].toInt()
            }
        }
    }

    override fun torrents(): Flowable<Torrent> {
        return service.torrents(
                configuration.endpoint,
                "",
                "main",
                "d.hash=",
                "d.name=",
                "d.state=",
                "d.down.rate=",
                "d.up.rate=",
                "d.peers_connected=",
                "d.peers_not_connected=",
                "d.bytes_done=",
                "d.up.total=",
                "d.size_bytes=",
                "d.left_bytes=",
                "d.creation_date=",
                "d.complete=",
                "d.is_active=",
                "d.is_hash_checking=",
                "d.base_path=",
                "d.base_filename=",
                "d.message=",
                "d.custom=addtime",
                "d.custom=seedingtime",
                "d.custom1=",
                "d.peers_complete=",
                "d.peers_accounted=")
                .flatten()
                .map { (hash, name, state, downloadRate, uploadRate, peersConnected, peersNotConnected, bytesDone, bytesUploaded, bytesTotal, bytesleft, timeCreated, isComplete, isActive, isHashChecking, basePath, baseFilename, errorMessage, timeAdded, timeFinished, label, seedersConnected, leechersConnected) ->
                    Torrent(
                            hash.hashCode().toLong(), hash, name,
                            torrentStatus(state, isComplete, isActive, isHashChecking),
                            basePath?.substring(0, basePath.indexOf(baseFilename.orEmpty())),
                            downloadRate.toInt(),
                            uploadRate.toInt(),
                            seedersConnected.toInt(),
                            (peersConnected + peersNotConnected).toInt(),
                            leechersConnected.toInt(),
                            (peersConnected + peersNotConnected).toInt(),
                            if (downloadRate > 0) bytesleft / downloadRate else null,
                            bytesDone, bytesUploaded, bytesTotal,
                            (bytesDone / bytesTotal).toFloat(),
                            null,
                            label,
                            torrentTimeAdded(timeAdded, timeCreated),
                            torrentTimeFinished(timeFinished),
                            errorMessage
                    )
                }
    }

    override fun files(torrent: Torrent): Flowable<TorrentFile> {
        return service.files(configuration.endpoint,
                torrent.uniqueId,
                "",
                "f.path=",
                "f.size_bytes=",
                "f.completed_chunks=",
                "f.size_chunks=",
                "f.priority=",
                "f.frozen_path=")
                .flatten()
                .zipWith(Flowable.range(0, Int.MAX_VALUE), BiFunction<FileSpec, Int, Pair<Int, FileSpec>> { file, index -> Pair(index, file) })
                .map { (index, file) ->
                    TorrentFile(
                            index.toString(),
                            file.pathName,
                            file.pathFull.substring(torrent.locationDir.orEmpty().length),
                            file.pathName,
                            file.size,
                            file.size * (file.chunksDone / file.chunksTotal),
                            fileStatus(file.priority))
                }
    }

    override fun details(torrent: Torrent): Single<TorrentDetails> {
        return service.trackers(configuration.endpoint, torrent.uniqueId, "", "t.url=")
                .flatten()
                .map { (url) -> url }
                .toList()
                .map { trackers -> TorrentDetails(trackers, listOfNotNull(torrent.error)) }
    }

    override fun start(torrent: Torrent): Single<Torrent> {
        return service.open(configuration.endpoint, torrent.uniqueId)
                .flatMap { service.start(configuration.endpoint, torrent.uniqueId) }
                .map { torrent.mimicStart() }
    }

    override fun stop(torrent: Torrent): Single<Torrent> {
        return service.stop(configuration.endpoint, torrent.uniqueId)
                .flatMap { service.close(configuration.endpoint, torrent.uniqueId) }
                .map { torrent.mimicStop() }
    }

    override fun resume(torrent: Torrent): Single<Torrent> {
        return service.start(configuration.endpoint, torrent.uniqueId)
                .map { torrent.mimicResume() }
    }

    override fun pause(torrent: Torrent): Single<Torrent> {
        return service.stop(configuration.endpoint, torrent.uniqueId)
                .map { torrent.mimicPause() }
    }

    override fun addByUrl(url: String): Completable {
        return clientVersion().asVersionInt().flatMap { integer ->
            if (integer >= 904) {
                service.loadStart(configuration.endpoint, "", url)
            } else {
                service.loadStart(configuration.endpoint, url)
            }
        }.toCompletable()
    }

    override fun addByMagnet(magnet: String): Completable {
        return clientVersion().asVersionInt().flatMap { integer ->
            if (integer >= 904) {
                service.loadStart(configuration.endpoint, "", magnet)
            } else {
                service.loadStart(configuration.endpoint, magnet)
            }
        }.toCompletable()
    }

    override fun addByFile(file: InputStream): Completable {
        return clientVersion().asVersionInt().flatMap { integer ->
            val bytes = file.readBytes()
            val size = Math.max(bytes.size, xmlrpcSizeMinimum) + xmlrpcSizePadding
            if (integer >= 904) {
                service.networkSizeLimitSet(configuration.endpoint, "", size)
                        .flatMap { service.loadRawStart(configuration.endpoint, "", bytes) }
            } else {
                service.networkSizeLimitSet(configuration.endpoint, size)
                        .flatMap { service.loadRawStart(configuration.endpoint, bytes) }
            }
        }.toCompletable()
    }

    private fun torrentStatus(state: Long, complete: Long, active: Long, checking: Long): TorrentStatus {
        if (state == 0L) {
            return TorrentStatus.QUEUED
        } else if (active == 1L) {
            if (complete == 1L) {
                return TorrentStatus.SEEDING
            } else {
                return TorrentStatus.DOWNLOADING
            }
        } else if (checking == 1L) {
            return TorrentStatus.CHECKING
        } else {
            return TorrentStatus.PAUSED
        }
    }

    private fun fileStatus(state: Long): Priority =
            when (state) {
                0L -> Priority.OFF
                2L -> Priority.HIGH
                else -> Priority.NORMAL
            }

    private fun torrentTimeAdded(timeAdded: String?, timeCreated: Long): Date =
            if (timeAdded.isNullOrBlank()) Date(timeCreated * 1000L) else Date(timeAdded!!.trim().toLong() * 1000L)

    private fun torrentTimeFinished(timeFinished: String?): Date? =
            if (timeFinished.isNullOrBlank()) null else Date(timeFinished!!.trim().toLong() * 1000L)

}
